### DISCUSSION TITLE
Needed Skeleton Modifiers

### DISCUSSION THREAD

**Discussion by SlashScreen** [godotengine/godot-proposals#9885] [General Discussions]
I've been working on some skeleton modifiers for 4.3, I've been prototyping them in GDScript, but later I would like to port them to GDExtension and put it on the asset library.

https://github.com/godotengine/godot-proposals/assets/12375239/4891a906-5a65-4222-b989-1d21b9e84a85

I've been compiling a list of modifiers to work on, but I'd like your comments.
- Look At (done) (I think look at multiple can just be multiple look at modifiers with varying levels of influence)
- Match Rotation (Done, but a bit buggy)
- Constant Transform Multiplier (done)
- Damping?
- Match Position? (Would this be useful for bones?)
- Match Scale?
- Match Property
- Limit Rotation
- Limit Position?

cc @fire @lyuma

**Comment by Broken1334** [comment]
Unity constraints, for reference.
- Aim: Rotate the constrained GameObject to face the linked GameObject.
- Look At: Rotate the constrained GameObject to the linked GameObject (simplified Aim Constraint).
- Parent: Move and rotate the constrained GameObject with the linked GameObject.
- Position: Move the constrained GameObject like the linked GameObject.
- Rotation: Rotate the constrained GameObject like the linked GameObject.
- Scale: Scale the constrained GameObject like the linked GameObject.

https://docs.unity3d.com/Manual/Constraints.html

Pretty much the same as in Maya.

**Comment by TokageItLab** [comment]
Here is what I would like to implement in the Core:
- [x] RetargetModifier3D
  - https://github.com/godotengine/godot-proposals/issues/3379#issuecomment-2306298660
  - https://github.com/godotengine/godot/pull/97824
- [x] TransformConstraint3D & Aim(Dump/Track/LockedTrack)Constraint3D
  - Based on Blender Constraint modifier (it can handle/unify/migrate almost all of those listed above)
  - https://docs.blender.org/manual/en/latest/animation/constraints/transform/transformation.html
  - https://github.com/godotengine/godot/pull/100984
- [x] SpringBone3D
  - Based on VRM1 but some extended options (e.g. freeze axis, angle speed limitation) are needed
  - https://github.com/vrm-c/vrm-specification/tree/master/specification/VRMC_springBone-1.0
  - https://github.com/godotengine/godot/pull/101409
- [ ] ~BoneExpander3D~ ~Superseded by "Add cancel bone" option (which is constrained by parent bone with inverse transform)~ BoneExpander3D
  - To solve https://github.com/godotengine/godot-proposals/issues/4190 simply
  - Constraint is available for workarounds, ~but this provides simple and intuitive scaling without changing the bone structure~ We found through experimentation that approach without changing the bone structure is not worked, because bone pose does not allow local shearing. And I and @lyuma agreed that a cancel bone by Constraint is the better solution
    - Experimental implementation have been sent as https://github.com/godotengine/godot/pull/111815
- [x] LookAtModifier3D
  - Needs to angle limitation and damping; Since when using for eye movement, it is desirable to have attenuation where the influence is not linear, but rather the degree of influence decreases as one moves outward
  - https://github.com/godotengine/godot/pull/98446
- [x] TwoBoneIK3D
  - Pole axis and angle limitation is needed
  - https://github.com/godotengine/godot/pull/110120
- [x] ChainedBoneIK3D(FABRIK/CCDIK)
  - Replace SkeletonIK3D
  - https://github.com/godotengine/godot/pull/110120
- [ ] FootIK3D
  - TwoBoneIK3D + ankle with a collider as the bottom plane of the foot
  - @Remi123 is working on?
- [ ] FullBodyIK3D
  - @fire is working on

**Reply by fire** [reply]
We discussed this list of modifiers at the Animation meeting today.

**Reply by SlashScreen** [reply]
Nice

**Reply by TokageItLab** [reply]
RetargetModifier PR https://github.com/godotengine/godot/pull/97824

**Reply by TokageItLab** [reply]
I sent a PR for TwoBoneIK and ChainIK (plus SplineIK) in https://github.com/godotengine/godot/pull/110120.

FootIK will likely be a combination of nodes like TwoBoneIK and Raycast. Therefore, I expect it will be provided later as an addon or tutorial that simplifies setting up these components, rather than as a single IK class.

Regarding FullBodyIK, while there is a module called [RenIK](https://github.com/V-Sekai/renik), considering its versatility, it might take some time to integrate it into the Godot core.

**Reply by TokageItLab** [reply]
I came up with an idea for how to handle BoneExpander although it is experimental, so I sent https://github.com/godotengine/godot/pull/111815.

**Comment by Broken1334** [comment]
And I'd like to see #9860 as future feature. It looks quite complex, but I hope still possible to implement.

**Reply by SlashScreen** [reply]
I don't really understand what this is.

**Reply by lyuma** [reply]
I was lucky to see the God of War Ragnarok talk at GDC, so I have some idea of the pieces involved.
It was an excellent talk, and they covered tons of interesting human anatomy as well as the depths of the vertex buffers they used for skin deformation.

They were awesome and uploaded the slides from their talk, so feel free to look through it:
https://sms.playstation.com/media/documents/GOWR_Tenghao_Wang_JointBasedSkin_GDC23.pdf

I think a short summary is doing a more physically accurate model of how muscles move, rather than merely thinking in terms of bones. Those "muscle" deformation bones will need to be driven from the actual animation bones, and that's where the SkeletonModifier3D would likely come into play.

**Reply by Broken1334** [reply]
> I was lucky to see the God of War Ragnarok talk at GDC, so I have some idea of the pieces involved.

Yes, this paper is great. 
They created an anatomy-based rig, but something more simple (like Metahuman) will work nice too.

> Those "muscle" deformation bones will need to be driven from the actual animation bones

Right, there is a driver joint (for example, lowerarm_r) and driven joints (2-4 joints for deformation correction), additionally you should create several corrective poses (rotate lowerarm_r to 20, 40, 60, 90 and move helper joints for good deformation in each pose), after that RBF solver can smoothly blend the helper joints position based on the driver joint orientation.

**Reply by fire** [reply]
Some adjacent work on cage deformer.
 
https://github.com/user-attachments/assets/ed555c28-ee9c-4a86-8a41-c0e95e498bf7

Investigated https://github.com/V-Sekai/unity-mesh-deform-direct-delta-mush skinning

**Comment by SlashScreen** [comment]
Alright, here's a sketch implementation of a few modifiers. 
https://github.com/SlashScreen/godot_constraints

**Comment by Broken1334** [comment]
Control Rig - how to improve character's deformation [Krzysiek Jedrzejek, CD Projekt RED]
https://www.youtube.com/watch?v=zXaSy4IcYAg

**Reply by Saul2022** [reply]
Wouldnt fullbody ik solve this or it's not the same ?
